internal/k8s/registry.go

```diff
@@ -49,66 +54,119 @@ func newRegistryAsync(env Env, core apiv1.CoreV1Interface, runtimeSource Runtime
+	if runtime == container.RuntimeDocker {
+		return container.Registry{}
+	}
 
-		// If Microk8s is using the docker runtime, we can just use the microk8s docker daemon
-		// instead of the registry.
-		runtime := r.runtimeSource.Runtime(ctx)
-		if runtime == container.RuntimeDocker {
-			return
+	// Microk8s might have a registry enabled.
+	// https://microk8s.io/docs/working
+	svc, err := r.core.Services(microk8sRegistryNamespace).Get(microk8sRegistryName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Get(ctx).Warnf("You are running microk8s without a local image registry.\n" +
+				"Run: `sudo microk8s.enable registry`\n" +
+				"Tilt will use the local registry to speed up builds")
+		} else {
+			logger.Get(ctx).Debugf("Error fetching services: %v", err)
 		}
+		return container.Registry{}
+	}
+
+	portSpecs := svc.Spec.Ports
+	if len(portSpecs) == 0 {
+		return container.Registry{}
+	}
+
+	// Check to make sure localhost resolves to an IPv4 address. If it doesn't,
+	// then we won't be able to connect to the registry. See:
+	// https://github.com/windmilleng/tilt/issues/2369
+	ips, err := net.LookupIP("localhost")
+	if err != nil || len(ips) == 0 || ips[0].To4() == nil {
+		logger.Get(ctx).Warnf("Your /etc/hosts is resolving localhost to ::1 (IPv6).\n" +
+			"This breaks the microk8s image registry.\n" +
+			"Please fix your /etc/hosts to default to IPv4. This will make image pushes much faster.")
+		return container.Registry{}
+	}
+
+	portSpec := portSpecs[0]
+	host := fmt.Sprintf("localhost:%d", portSpec.NodePort)
+	reg, err := container.NewRegistry(host)
+	if err != nil {
+		logger.Get(ctx).Warnf("Error validating private registry host %q: %v", host, err)
+		return container.Registry{}
+	}
+
+	return reg
+}
+
+func (r *registryAsync) inferRegistryFromTiltNodeLabels(ctx context.Context) container.Registry {
+	nodeList, err := r.core.Nodes().List(metav1.ListOptions{Limit: 1})
+	if err != nil || len(nodeList.Items) == 0 {
+		return container.Registry{}
+	}
+
+	node := nodeList.Items[0]
+	labels := node.Labels
 
-		// Microk8s might have a registry enabled.
-		// https://microk8s.io/docs/working
-		svc, err := r.core.Services(microk8sRegistryNamespace).Get(microk8sRegistryName, metav1.GetOptions{})
+	fromLocal := ""
+	fromCluster := ""
+
+	registryHostnameFromLocal := labels[labelRegistryHostnameFromLocal]
+	registryPortFromLocal := labels[labelRegistryPortFromLocal]
+	if registryHostnameFromLocal != "" && registryPortFromLocal != "" {
+		fromLocal = fmt.Sprintf("%s:%s", registryHostnameFromLocal, registryPortFromLocal)
+	}
+
+	registryHostnameFromCluster := labels[labelRegistryHostnameFromCluster]
+	registryPortFromCluster := labels[labelRegistryPortFromCluster]
+	if registryHostnameFromCluster != "" && registryPortFromCluster != "" {
+		fromCluster = fmt.Sprintf("%s:%s", registryHostnameFromCluster, registryPortFromCluster)
+	}
+
+	if fromLocal != "" && fromCluster != "" {
```

Comment:
You can pass `""` as the 2nd arg to `NewRegistryWithHostFromCluster`, that might be easier to follow than these conditionals

i.e.
```
var fromCluster
if registryHostnameFromCluster != "" && registryPortFromCluster != "" {
    fromCluster = fmt.Sprintf("%s:%s", registryHostnameFromCluster, registryPortFromCluster)
}
eg, err := container.NewRegistryWithHostFromCluster(fromLocal, fromCluster)
...
```

Reply:
done